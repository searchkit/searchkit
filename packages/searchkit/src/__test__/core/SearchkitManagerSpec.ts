import * as _ from 'lodash'
import {
  SearchkitManager,
  AccessorManager,
  AxiosESTransport,
  ImmutableQuery,
  createHistoryInstance,
  PageSizeAccessor,
  SearchRequest,
  EventEmitter,
  QueryAccessor,
  MockESTransport
} from '../../'

describe('SearchkitManager', () => {
  beforeEach(() => {
    this.host = 'http://localhost:9200'
    spyOn(SearchkitManager.prototype, 'runInitialSearch').and.callThrough()
    this.searchkit = new SearchkitManager(this.host, {
      useHistory: false,
      httpHeaders: {
        'Content-Type': 'application/json'
      },
      basicAuth: 'key:val',
      searchUrlPath: '/search',
      searchOnLoad: false
    })
    spyOn(SearchRequest.prototype, 'run').and.returnValue(Promise.resolve())

    this.searchkit.setupListeners()
    this.emitterSpy = jasmine.createSpy('emitter')
    this.searchkit.emitter.addListener(this.emitterSpy)
    this.accessors = this.searchkit.accessors
    expect(this.searchkit.transport.options.searchUrlPath).toBe('/search')
    expect(SearchkitManager.prototype.runInitialSearch).toHaveBeenCalled()
  })

  it('constructor()', () => {
    expect(this.searchkit.host).toBe(this.host)
    expect(this.searchkit.accessors).toEqual(jasmine.any(AccessorManager))
    expect(this.searchkit.registrationCompleted).toEqual(jasmine.any(Promise))
    expect(this.searchkit.translateFunction).toEqual(jasmine.any(Function))
    expect(this.searchkit.transport).toEqual(jasmine.any(AxiosESTransport))
    expect(this.searchkit.transport.options.headers).toEqual(
      jasmine.objectContaining({
        'Content-Type': 'application/json'
      })
    )
    expect(this.searchkit.transport.axios.defaults.auth).toEqual(
      jasmine.objectContaining({
        username: 'key',
        password: 'val'
      })
    )
    expect(this.searchkit.query).toEqual(new ImmutableQuery())
    expect(this.searchkit.emitter).toEqual(jasmine.any(EventEmitter))
    expect(this.searchkit.options.searchOnLoad).toBe(false)
    expect(this.searchkit.initialLoading).toBe(true)
    expect(this.searchkit.results).toEqual(undefined)
    expect(this.searchkit.state).toEqual({})
    expect(this.searchkit.options.withCredentials).toBeFalsy()
    //check queryProcessor is an identity function
    expect(this.searchkit.queryProcessor('query')).toBe('query')
  })

  it('constructor - withCredentials', () => {
    this.searchkit = new SearchkitManager(this.host, {
      useHistory: false,
      httpHeaders: {
        'Content-Type': 'application/json'
      },
      withCredentials: true,
      searchUrlPath: '/search',
      searchOnLoad: false
    })
    expect(this.searchkit.options.withCredentials).toEqual(true)
  })

  it('SearchkitManager.mock()', () => {
    const searchkit = SearchkitManager.mock()
    expect(searchkit.host).toBe('/')
    expect(searchkit.options.useHistory).toBe(false)
    expect(searchkit.options.transport).toEqual(jasmine.any(MockESTransport))
  })

  it('defaultSize', () => {
    let searchkit = SearchkitManager.mock()
    expect(searchkit.buildQuery().getSize()).toEqual(20)

    searchkit = SearchkitManager.mock({ defaultSize: 12 })
    expect(searchkit.buildQuery().getSize()).toEqual(12)
    expect(searchkit.getAccessorByType(PageSizeAccessor).defaultSize).toEqual(12)
  })

  it('addAccessor(), removeAddAccessor()', () => {
    this.searchkit.accessors = new AccessorManager()
    const accessor = new PageSizeAccessor(10)
    this.searchkit.addAccessor(accessor)
    expect(this.searchkit.accessors.accessors).toEqual([accessor])
    this.searchkit.removeAccessor(accessor)
    expect(this.searchkit.accessors.accessors).toEqual([])
  })

  it('addDefaultQuery()', () => {
    const queryFn = (query) => query.setSize(11)
    this.searchkit.addDefaultQuery(queryFn)
    expect(this.searchkit.buildQuery().getSize()).toBe(11)
  })

  it('translate()', () => {
    spyOn(this.searchkit, 'translateFunction').and.callThrough()
    expect(this.searchkit.translate('foo')).toBe(undefined)
    expect(this.searchkit.translateFunction).toHaveBeenCalledWith('foo')
  })

  it('buildQuery()', () => {
    this.searchkit.accessors = new AccessorManager()
    const defaultQueryFn = (query) => query.setFrom(20)
    this.searchkit.addDefaultQuery(defaultQueryFn)
    this.searchkit.addAccessor(new PageSizeAccessor(10))
    const query = this.searchkit.buildQuery()
    expect(query.getSize()).toBe(10)
    expect(query.getFrom()).toBe(20)
  })

  it('resetState()', () => {
    spyOn(this.accessors, 'resetState')
    this.searchkit.resetState()
    expect(this.accessors.resetState).toHaveBeenCalled()
  })

  it('listenToHistory()', (done) => {
    const history = createHistoryInstance()
    history.push(window.location.pathname + '?q=foo')
    SearchkitManager.prototype.unlistenHistory = jasmine.createSpy('unlisten')
    const searchkit = new SearchkitManager('/', {
      useHistory: true
    })
    searchkit.setupListeners()
    expect(SearchkitManager.prototype.unlistenHistory).toHaveBeenCalled()
    spyOn(searchkit.accessors, 'setState')
    spyOn(searchkit, '_search')
    searchkit.completeRegistration()
    setTimeout(() => {
      expect(searchkit._search).toHaveBeenCalled()
      expect(searchkit.accessors.setState).toHaveBeenCalledWith({ q: 'foo' })
      searchkit.unlistenHistory()
      done()
    }, 0)
  })

  it('listenToHistory() - searchOnLoad false', (done) => {
    const history = createHistoryInstance()
    history.push(window.location.pathname + '?q=foo-previous')
    history.push(window.location.pathname + '?q=foo-now')
    setTimeout(() => {
      const searchkit = new SearchkitManager('/', {
        useHistory: true,
        searchOnLoad: false
      })
      spyOn(searchkit.accessors, 'setState')
      spyOn(searchkit, 'searchFromUrlQuery')
      spyOn(searchkit, '_search')
      searchkit.setupListeners()
      searchkit.completeRegistration()
      setTimeout(() => {
        history.goBack()
        setTimeout(() => {
          expect(searchkit.searchFromUrlQuery).toHaveBeenCalledWith('?q=foo-previous')
          searchkit.unlistenHistory()
          done()
        }, 10)
      }, 0)
    }, 0)
  })

  it('listenToHistory() - handle error', (done) => {
    const history = createHistoryInstance()
    history.push(window.location.pathname + '?q=foo')
    const searchkit = new SearchkitManager('/', {
      useHistory: true
    })
    searchkit.setupListeners()

    const error = new Error('oh no')
    searchkit.searchFromUrlQuery = () => {
      throw error
    }
    spyOn(console, 'error')
    const consoleErrorSpy = console.error as jasmine.Spy
    searchkit.completeRegistration()
    setTimeout(() => {
      const consoleErrorCallArg = consoleErrorSpy.calls.argsFor(0)[0]
      expect(consoleErrorCallArg).toBe(error)
      searchkit.unlistenHistory()
      done()
    }, 0)
  })

  it('performSearch()', () => {
    const searchkit = new SearchkitManager('/', {
      useHistory: true
    })
    searchkit.setupListeners()
    searchkit.state = {
      q: 'foo'
    }
    spyOn(searchkit.accessors, 'notifyStateChange')
    spyOn(searchkit, '_search').and.returnValue(true)
    spyOn(searchkit.history, 'push')
    expect(searchkit.performSearch()).toEqual(true)
    expect(searchkit.history.push).toHaveBeenCalledWith('/context.html?q=foo')
    expect(searchkit.accessors.notifyStateChange).toHaveBeenCalledWith(searchkit.state)
    searchkit.unlistenHistory()
  })

  it('run initial search', (done) => {
    const searchkit = new SearchkitManager(this.host, {
      useHistory: false,
      searchOnLoad: false
    })
    spyOn(searchkit, '_search')
    expect(SearchkitManager.prototype.runInitialSearch).toHaveBeenCalled()
    searchkit.completeRegistration()
    setTimeout(() => {
      expect(searchkit._search).not.toHaveBeenCalled()
      searchkit.options.searchOnLoad = true
      searchkit.runInitialSearch()
      setTimeout(() => {
        expect(searchkit._search).toHaveBeenCalled()
        done()
      })
    })
  })

  it('performSearch() - same state + replaceState', () => {
    const searchkit = new SearchkitManager('/', {
      useHistory: true
    })
    searchkit.setupListeners()
    searchkit.state = {
      q: 'foo'
    }
    searchkit.accessors.getState = () => ({ q: 'foo' })
    spyOn(searchkit.accessors, 'notifyStateChange')
    spyOn(searchkit, '_search').and.returnValue(true)
    spyOn(searchkit.history, 'replace')
    expect(searchkit.performSearch(true)).toEqual(true)
    expect(searchkit.history.replace).toHaveBeenCalled()
    expect(searchkit.accessors.notifyStateChange).not.toHaveBeenCalled()
    searchkit.unlistenHistory()
    searchkit.state = { q: 'bar' }
    searchkit.performSearch(true, false)
    expect(searchkit.accessors.notifyStateChange).not.toHaveBeenCalled()
    searchkit.performSearch(true, true)
    expect(searchkit.accessors.notifyStateChange).toHaveBeenCalled()
  })

  it('search()', () => {
    spyOn(this.searchkit, 'performSearch').and.returnValue(true)
    expect(this.searchkit.search()).toEqual(true)
    expect(this.searchkit.performSearch).toHaveBeenCalled()
  })

  it('_search()', async () => {
    this.searchkit.setQueryProcessor((query) => {
      query.source = true
      return query
    })
    const initialSearchRequest = (this.searchkit.currentSearchRequest = new SearchRequest(
      this.host,
      null,
      this.searchkit
    ))

    this.searchkit.results = {}
    const resultsObject = await this.searchkit._search()

    expect(resultsObject).toEqual({
      results: {},
      state: {}
    })

    expect(initialSearchRequest.active).toBe(false)
    expect(this.searchkit.currentSearchRequest.transport.host).toBe(this.host)
    expect(this.searchkit.currentSearchRequest.query).toEqual({
      size: 20,
      source: true
    })
    expect(this.searchkit.currentSearchRequest.run).toHaveBeenCalled()
    expect(this.searchkit.loading).toBe(true)
  })

  it('_search() should not search with same query', async () => {
    this.searchkit.query = new ImmutableQuery().setSize(20).setSort([{ created: 'desc' }])
    this.searchkit.buildQuery = () =>
      new ImmutableQuery().setSize(20).setSort([{ created: 'desc' }])
    this.searchkit.results = {}
    this.searchkit._search()

    expect(SearchRequest.prototype.run).not.toHaveBeenCalled()

    delete this.searchkit.results
    let searchResult = await this.searchkit._search()
    expect(searchResult).toEqual({ results: undefined, state: {} })
    expect(SearchRequest.prototype.run).toHaveBeenCalled()
    this.searchkit.query = new ImmutableQuery().setSize(21)
    this.searchkit.results = {}
    searchResult = await this.searchkit._search()
    expect(searchResult).toEqual({ results: {}, state: {} })
    expect(SearchRequest.prototype.run).toHaveBeenCalled()
  })

  it('_search() should use shouldSearch check', async () => {
    this.searchkit.buildQuery = () =>
      new ImmutableQuery().setSize(20).setSort([{ created: 'desc' }])
    this.searchkit.results = {}
    this.searchkit.shouldPerformSearch = (query) => !!query.getQueryString()
    let searchResult = await this.searchkit._search()
    expect(searchResult).toEqual({ results: {}, state: {} })
    expect(SearchRequest.prototype.run).not.toHaveBeenCalled()

    this.searchkit.buildQuery = () =>
      new ImmutableQuery()
        .setSize(20)
        .setSort([{ created: 'desc' }])
        .setQueryString('somequery')

    searchResult = await this.searchkit._search()
    expect(searchResult).toEqual({ results: {}, state: {} })
    expect(SearchRequest.prototype.run).toHaveBeenCalled()
  })

  it('reloadSearch()', () => {
    this.searchkit.query = new ImmutableQuery().setSize(20).setSort([{ created: 'desc' }])
    this.searchkit.buildQuery = () =>
      new ImmutableQuery().setSize(20).setSort([{ created: 'desc' }])
    this.searchkit.results = {}
    this.searchkit._search()
    expect(SearchRequest.prototype.run).not.toHaveBeenCalled()
    this.searchkit.reloadSearch()
    expect(SearchRequest.prototype.run).toHaveBeenCalled()
  })

  it('setResults()', () => {
    spyOn(this.accessors, 'setResults')
    spyOn(this.searchkit, 'onResponseChange')
    expect(this.searchkit.results).toBe(undefined)
    const resultsSpy = jasmine.createSpy('results')
    const removalFn = this.searchkit.addResultsListener(resultsSpy)
    expect(removalFn).toEqual(jasmine.any(Function))
    this.searchkit.setResults('foo')
    expect(this.searchkit.results).toBe('foo')
    expect(this.accessors.setResults).toHaveBeenCalledWith('foo')
    expect(this.searchkit.onResponseChange).toHaveBeenCalled()
    expect(resultsSpy).toHaveBeenCalledWith('foo')
  })

  it('setResults() - error', () => {
    spyOn(this.searchkit, 'onResponseChange')
    spyOn(this.accessors, 'setResults')
    spyOn(console, 'error')
    expect(this.searchkit.results).toBe(undefined)
    const error = new Error('oh no')
    this.searchkit.setError(error)
    expect(this.searchkit.error).toBe(error)
    expect(console.error).toHaveBeenCalledWith(error)
    expect(this.searchkit.results).toBe(null)
    expect(this.accessors.setResults).toHaveBeenCalledWith(null)
    expect(this.searchkit.onResponseChange).toHaveBeenCalled()
  })

  it('setResults() - change detection', () => {
    spyOn(this.accessors, 'setResults')
    spyOn(this.searchkit, 'onResponseChange')
    const results = {
      hits: {
        total: 2,
        hits: [
          { _id: 1, _source: { title: 'Doc1' } },
          { _id: 2, _source: { title: 'Doc2' } }
        ]
      }
    }
    this.searchkit.setResults(_.cloneDeep(results))
    expect(this.searchkit.results.hits.ids).toBe('1,2')
    expect(this.searchkit.results.hits.hasChanged).toBe(true)
    expect(this.searchkit.hasHitsChanged()).toBe(true)

    this.searchkit.setResults(_.cloneDeep(results))
    expect(this.searchkit.hasHitsChanged()).toBe(false)
    results.hits.hits.push({ _id: 3, _source: { title: 'Doc3' } })
    this.searchkit.setResults(_.cloneDeep(results))
    expect(this.searchkit.results.hits.ids).toBe('1,2,3')
    expect(this.searchkit.hasHitsChanged()).toBe(true)
  })
  it('guid()', () => {
    expect(this.searchkit.guid('foo')).toEqual('foo1')
    expect(this.searchkit.guid('bar')).toEqual('bar2')
  })

  it('getHits()', () => {
    expect(this.searchkit.getHits()).toEqual([])
    this.searchkit.results = {
      hits: {
        hits: [1, 2, 3, 4]
      }
    }
    expect(this.searchkit.getHits()).toEqual([1, 2, 3, 4])
  })

  it('getHitsCount(), hasHits()', () => {
    expect(this.searchkit.getHitsCount()).toEqual(0)
    expect(this.searchkit.hasHits()).toBe(false)
    this.searchkit.results = {
      hits: {
        total: 99
      },
      took: 1
    }
    expect(this.searchkit.getHitsCount()).toBe(99)
    expect(this.searchkit.getTime()).toBe(1)
    expect(this.searchkit.hasHits()).toBe(true)
  })

  it('getHitsCount(), hasHits() ES7', () => {
    expect(this.searchkit.getHitsCount()).toEqual(0)
    expect(this.searchkit.hasHits()).toBe(false)
    this.searchkit.results = {
      hits: {
        total: {
          relation: 'eq',
          value: 99
        }
      },
      took: 1
    }
    expect(this.searchkit.getHitsCount()).toBe(99)
    expect(this.searchkit.getTime()).toBe(1)
    expect(this.searchkit.hasHits()).toBe(true)
  })

  it('getQueryAccessor()', () => {
    const queryAccessor = new QueryAccessor('q')
    this.searchkit.addAccessor(queryAccessor)
    expect(this.searchkit.getQueryAccessor()).toBe(queryAccessor)
  })

  it('getAccessorsByType(), getAccessorByType()', () => {
    const queryAccessor = new QueryAccessor('q')
    this.searchkit.addAccessor(queryAccessor)
    expect(this.searchkit.getAccessorsByType(QueryAccessor)).toEqual([queryAccessor])
    expect(this.searchkit.getAccessorByType(QueryAccessor)).toEqual(queryAccessor)
  })

  it('onResponseChange()', () => {
    this.searchkit.loading = true
    this.searchkit.initialLoading = true
    this.searchkit.onResponseChange()
    expect(this.searchkit.loading).toBe(false)
    expect(this.searchkit.initialLoading).toBe(false)
    expect(this.emitterSpy).toHaveBeenCalled()
  })
})
