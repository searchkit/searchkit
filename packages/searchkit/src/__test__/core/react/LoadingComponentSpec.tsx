import * as React from "react";

import {
  SearchkitManager,
  LoadingComponent
} from "../../../"

import {mount} from "enzyme";

describe("SearchkitProvider", ()=> {

  beforeEach(()=> {
    this.searchkit = SearchkitManager.mock()
    this.wrapper = mount(
      <LoadingComponent searchkit={this.searchkit}>
        <p>loading...</p>
      </LoadingComponent>
    )
  })

  it("display nothing when not loading", ()=> {
    expect(this.wrapper).toMatchSnapshot()
  })

  it("display children when loading", ()=> {
    this.searchkit.loading = true
    this.wrapper.instance().forceUpdate()
    expect(this.wrapper).toMatchSnapshot()
  })


})
