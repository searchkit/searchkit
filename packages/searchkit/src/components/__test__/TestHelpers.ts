export const fastClick = (el)=>{
  el.simulate("click")
}