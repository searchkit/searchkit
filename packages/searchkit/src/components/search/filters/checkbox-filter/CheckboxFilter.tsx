import * as PropTypes from "prop-types";

import {
  SearchkitComponent,
  SearchkitComponentProps,
  CheckboxFilterAccessor,
  RenderComponentType,
  renderComponent
} from "../../../../core";

import {
  Panel, CheckboxItemList
} from "../../../ui"

const defaults = require("lodash/defaults")

export interface CheckboxFilterProps extends SearchkitComponentProps {
  id: string
  filter: any
  title: string
  label: string
  containerComponent?: RenderComponentType<any>
  listComponent?: RenderComponentType<any>
  showCount?: boolean
}

export class CheckboxFilter extends SearchkitComponent<CheckboxFilterProps, any> {
  accessor: CheckboxFilterAccessor

  static propTypes = defaults({
    id: PropTypes.string.isRequired,
    title: PropTypes.string.isRequired,
    label: PropTypes.string.isRequired,
    filter: PropTypes.object.isRequired,
    translations: SearchkitComponent.translationsPropType(
        CheckboxFilterAccessor.translations
    ),
    showCount: PropTypes.bool,
  }, SearchkitComponent.propTypes)

  static defaultProps = {
    listComponent: CheckboxItemList,
    containerComponent: Panel,
    collapsable: false,
    showCount: true
  }

  constructor(props){
    super(props)
    this.toggleFilter = this.toggleFilter.bind(this)
  }

  defineAccessor() {
    const { id, title, translations, label, filter } = this.props;
    return new CheckboxFilterAccessor(id, {
      id, title, label, translations, filter
    })
  }

  toggleFilter() {
    this.accessor.state = this.accessor.state.create(!this.accessor.state.getValue())
    this.searchkit.performSearch()
  }

  setFilters(keys) {
    this.accessor.state = this.accessor.state.setValue(keys.length > 0)
    this.searchkit.performSearch()
  }

  getSelectedItems() {
    if (this.accessor.state.getValue()) {
      return [this.props.label]
    } else {
      return []
    }
  }

  render(): React.ReactElement<any> {
    if(!this.accessor) return null;
    const { listComponent, containerComponent, showCount, title, id, label } = this.props

    const disabled =  (this.searchkit.getHitsCount() == 0) && !this.accessor.state.getValue()

    return renderComponent(containerComponent, {
      title,
      className: id ? `filter--${id}` : undefined,
      disabled
    },
      renderComponent(listComponent, {
        items: [{ key: label, doc_count: this.accessor.getDocCount() }],
        selectedItems: this.getSelectedItems(),
        toggleItem: this.toggleFilter,
        setItems: this.setFilters.bind(this),
        showCount
      })
    );
  }

}
