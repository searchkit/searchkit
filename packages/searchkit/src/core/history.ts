import { createBrowserHistory, createMemoryHistory,  History } from 'history'
const qs = require("qs")

export const encodeObjUrl = (obj) => {
  return qs.stringify(obj, { encode: true, encodeValuesOnly: true })
}

export const decodeObjString = (str) => {
  return qs.parse(str)
}

export const supportsHistory = ()=> {
  return !!window.history
}

export const createHistoryInstance = function():History{
  return supportsHistory() ? createBrowserHistory() : createMemoryHistory()
}
