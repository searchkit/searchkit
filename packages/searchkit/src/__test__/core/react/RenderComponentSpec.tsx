import * as React from "react";
import * as PropTypes from "prop-types";
import {mount} from "enzyme";
import {
  fastClick
} from "../../../components/__test__/TestHelpers"

import {Panel} from "../../../components"

import {
  renderComponent,
  RenderComponentType,
  RenderComponentPropType
} from "../../../"


describe("RenderComponent", ()=> {

  beforeEach(()=> {
    this.SubPanel = class SubPanel extends Panel {
    }
    this.SubPanel.defaultProps.title = "SubPanel"
    this.SubPanelElement = <Panel title="PanelElement"/>

    this.PanelReactClass = class PanelReactClass extends React.Component {
      static contextTypes = {
        color:PropTypes.string
      }
      render(){
        return (<Panel title={"PanelReactClass " + this.context.color} {...this.props}>
          {this.props.children}
        </Panel>)
      }
    }



    this.PanelFunction = (props, context)=> {
      return (
        <Panel title={"PanelFunction " + context.color}>
          {props.children}
        </Panel>
      )
    }
    this.PanelFunction.contextTypes = {
      color:PropTypes.string
    }

    class Provider extends React.Component<any, any>{
      static childContextTypes = {
        color: PropTypes.string
      }
      getChildContext(){
        return {color:"purple"}
      }
      render(){
        return this.props.children
      }
    }

    this.mount = (component, props={})=> {
      this.wrapper = mount(
        <Provider>
          {renderComponent(
            component, props,
            <p>content..</p>
          )}
        </Provider>
      )
    }

  })

  it("React.Component class", ()=> {
    this.mount(this.SubPanel)
    expect(this.wrapper).toMatchSnapshot()

  })

  it("React Element", ()=> {
    this.mount(this.SubPanelElement)
    expect(this.wrapper).toMatchSnapshot()
  })

  it("React Class", ()=> {
    this.mount(this.PanelReactClass)
    expect(this.wrapper).toMatchSnapshot()
  })

  it("Render function", ()=> {
    this.mount(this.PanelFunction)
    expect(this.wrapper).toMatchSnapshot()
  })

  it("Invalid component", ()=> {
    spyOn(console, "warn")
    try{
      this.mount(10)
    } catch (e){

    }
    expect(console.warn).toHaveBeenCalledWith(
      "Invalid component", 10
    )

  })

})
