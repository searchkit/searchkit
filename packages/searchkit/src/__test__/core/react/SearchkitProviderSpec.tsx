import * as React from "react";

import {
  SearchkitProvider,
  SearchkitManager,
  SearchkitComponent
} from "../../../"

import {mount} from "enzyme";

describe("SearchkitProvider", ()=> {

  beforeEach(()=> {
    this.searchkit = SearchkitManager.mock()
    class SomeComponent extends SearchkitComponent<any, any> {

      render(){
        return <h1>Hello</h1>
      }
    }
    this.wrapper = mount(
      <SearchkitProvider searchkit={this.searchkit}>
        <SomeComponent/>
      </SearchkitProvider>
    )
  })

  it("searchkit provider should work correctly", ()=> {
    expect(this.wrapper.html()).toBe("<h1>Hello</h1>")
    expect(this.wrapper.node.props.searchkit)
      .toBe(this.searchkit)
  })

  it("should call setupListeners()", ()=> {
    spyOn(this.searchkit, "setupListeners")
    expect(this.searchkit.setupListeners).not.toHaveBeenCalled()
    this.wrapper.node.componentWillMount()
    expect(this.searchkit.setupListeners).toHaveBeenCalled()
    this.searchkit.guidGenerator.counter = 10
    this.searchkit.unlistenHistory = jasmine.createSpy("unlisten")
    this.wrapper.node.componentWillUnmount()
    expect(this.searchkit.unlistenHistory).toHaveBeenCalled()
    expect(this.searchkit.guidGenerator.counter).toEqual(0)

  })


})
