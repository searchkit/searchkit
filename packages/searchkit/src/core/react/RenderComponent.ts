import * as React from 'react'
import * as PropTypes from "prop-types"

const omitBy = require("lodash/omitBy")
const isUndefined = require("lodash/isUndefined")

export type RenderFunction = (props?:any, children?:any) => Element
export type Element = React.ReactElement<any>
export type RenderComponentType<P> = React.ComponentClass<P> | React.ClassicComponentClass<P> | Element | RenderFunction | any

export const RenderComponentPropType = PropTypes.oneOfType([
  function(props:any, propName: string) {
     if(isUndefined(props[propName]) || (props[propName]["prototype"] instanceof React.Component)) {
       return null
     }
  },
  PropTypes.element,
  PropTypes.func,
])

export function renderComponent(component:RenderComponentType<any>, props={}, children=null): React.ReactElement<any> {
  let isReactComponent = (
    component["prototype"] instanceof React.Component ||
    (component["prototype"] && component["prototype"].isReactComponent) ||
    typeof component === 'function'
  )
  if (isReactComponent){
    return React.createElement(
      component as React.ComponentClass<any>,
      props, children
    )
  } else if (React.isValidElement(component)){
    return React.cloneElement(
      component as Element,
      omitBy(props, isUndefined), children
    );
  }
  console.warn("Invalid component", component)
  return null
}
