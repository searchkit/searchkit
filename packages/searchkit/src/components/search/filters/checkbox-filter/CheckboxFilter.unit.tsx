import * as React from "react";;
import {mount, render} from "enzyme";
import { fastClick} from "../../../__test__/TestHelpers"
import {CheckboxFilter} from "./CheckboxFilter";
import {SearchkitManager, Utils} from "../../../../core";
import {Toggle, ItemComponent} from "../../../ui";
import { TermQuery, CheckboxFilterAccessor } from "../../../../core";

import * as _ from "lodash"
import * as sinon from "sinon";

describe("CheckboxFilter tests", () => {
  this.createWrapper = (component) => {
    this.wrapper = mount(component)

    this.searchkit.setResults({
      hits:{
        hits:[{_id:1, title:1},{_id:2,title:2}],
        total:2
      },
      aggregations: {
        "test id1": {
          doc_count: 50
        }
      }
    })

    this.accessor = this.searchkit.getAccessorByType(CheckboxFilterAccessor)
  }

  beforeEach(() => {
    Utils.guidCounter = 0

    this.searchkit = SearchkitManager.mock()
    this.searchkit.translateFunction = (key) => {
      return {
        "test option 1": "test option 1 translated"
      }[key]
    }

    this.createWrapper(
      <CheckboxFilter
        id="test id" title="test title" label="test label"
        searchkit={this.searchkit}
        filter={TermQuery("test-field", "test-value")} />
    )
  });

  it('renders correctly', () => {
    expect(this.wrapper).toMatchSnapshot()
  });

  it('clicks options', () => {
    expect(this.accessor.state.getValue()).toEqual(null)
    let option = this.wrapper.find(".sk-item-list").children().at(0)
    expect(this.wrapper).toMatchSnapshot("option is not active")
    fastClick(option)
    expect(this.wrapper).toMatchSnapshot("option is active")
    expect(this.accessor.state.getValue()).toEqual(true)
    fastClick(option)
    expect(this.accessor.state.getValue()).toEqual(false) // Back to null ?
  })

  it("should configure accessor correctly", () => {
    expect(this.accessor.key).toBe("test id")
    let options = this.accessor.options

    expect(options).toEqual({
      "id": "test id",
      "title": "test title",
      "label": "test label",
      "translations": undefined,
      "filter": TermQuery("test-field", "test-value")
    })
  })

  it("can disable", () => {
    expect(this.wrapper).toMatchSnapshot("panel is not disabled")

    this.searchkit.setResults({
      hits:{ total:0 },
      aggregations: {
        "test id1": {
          doc_count: 50
        }
      }
    })
    expect(this.wrapper).toMatchSnapshot("panel is disabled")

    expect(this.accessor.state.getValue()).toEqual(null)
    let option = this.wrapper.find(".sk-item-list").children().at(0)
    fastClick(option)
    expect(this.accessor.state.getValue()).toEqual(true)

    expect(this.wrapper).toMatchSnapshot("panel is not disabled again")

  })
});
