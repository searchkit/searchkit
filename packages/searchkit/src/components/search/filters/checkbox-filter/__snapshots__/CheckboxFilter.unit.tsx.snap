// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CheckboxFilter tests renders correctly 1`] = `
<div class="sk-panel filter--test id">
  <div class="sk-panel__header">test title</div>
  <div class="sk-panel__content">
    <div data-qa="options" class="sk-item-list">
      <div class="sk-item-list-option sk-item-list__item" data-qa="option" data-key="test label"><input type="checkbox" data-qa="checkbox" readonly="" class="sk-item-list-option__checkbox">
        <div data-qa="label" class="sk-item-list-option__text">test label</div>
        <div data-qa="count" class="sk-item-list-option__count">50</div>
      </div>
    </div>
  </div>
</div>
`;

exports[`option is active 1`] = `
<div class="sk-panel filter--test id">
  <div class="sk-panel__header">test title</div>
  <div class="sk-panel__content">
    <div data-qa="options" class="sk-item-list">
      <div class="sk-item-list-option sk-item-list__item is-active" data-qa="option" data-key="test label"><input type="checkbox" data-qa="checkbox" readonly="" class="sk-item-list-option__checkbox is-active">
        <div data-qa="label" class="sk-item-list-option__text">test label</div>
        <div data-qa="count" class="sk-item-list-option__count">50</div>
      </div>
    </div>
  </div>
</div>
`;

exports[`option is not active 1`] = `
<div class="sk-panel filter--test id">
  <div class="sk-panel__header">test title</div>
  <div class="sk-panel__content">
    <div data-qa="options" class="sk-item-list">
      <div class="sk-item-list-option sk-item-list__item" data-qa="option" data-key="test label"><input type="checkbox" data-qa="checkbox" readonly="" class="sk-item-list-option__checkbox">
        <div data-qa="label" class="sk-item-list-option__text">test label</div>
        <div data-qa="count" class="sk-item-list-option__count">50</div>
      </div>
    </div>
  </div>
</div>
`;

exports[`panel is disabled 1`] = `
<div class="sk-panel filter--test id is-disabled">
  <div class="sk-panel__header is-disabled">test title</div>
  <div class="sk-panel__content is-disabled">
    <div data-qa="options" class="sk-item-list">
      <div class="sk-item-list-option sk-item-list__item" data-qa="option" data-key="test label"><input type="checkbox" data-qa="checkbox" readonly="" class="sk-item-list-option__checkbox">
        <div data-qa="label" class="sk-item-list-option__text">test label</div>
        <div data-qa="count" class="sk-item-list-option__count">50</div>
      </div>
    </div>
  </div>
</div>
`;

exports[`panel is not disabled 1`] = `
<div class="sk-panel filter--test id">
  <div class="sk-panel__header">test title</div>
  <div class="sk-panel__content">
    <div data-qa="options" class="sk-item-list">
      <div class="sk-item-list-option sk-item-list__item" data-qa="option" data-key="test label"><input type="checkbox" data-qa="checkbox" readonly="" class="sk-item-list-option__checkbox">
        <div data-qa="label" class="sk-item-list-option__text">test label</div>
        <div data-qa="count" class="sk-item-list-option__count">50</div>
      </div>
    </div>
  </div>
</div>
`;

exports[`panel is not disabled again 1`] = `
<div class="sk-panel filter--test id">
  <div class="sk-panel__header">test title</div>
  <div class="sk-panel__content">
    <div data-qa="options" class="sk-item-list">
      <div class="sk-item-list-option sk-item-list__item is-active" data-qa="option" data-key="test label"><input type="checkbox" data-qa="checkbox" readonly="" class="sk-item-list-option__checkbox is-active">
        <div data-qa="label" class="sk-item-list-option__text">test label</div>
        <div data-qa="count" class="sk-item-list-option__count">50</div>
      </div>
    </div>
  </div>
</div>
`;
